### DISCUSSION TITLE
Easier way to flatten an object with STJ serialization

### DISCUSSION THREAD

**Discussion by LeaFrock** [dotnet/runtime#77206] [Q&A] [answered]
Recently I meet a requirement that serializes a object with STJ. See the example below,

```csharp

    public ItemsData Item = new()
    {
         // Skip others' initialization

         Dynamics = new()
                {
                    { "prop1", "1" },
                    { "prop2", "2" },
                    { "prop3", "3" },
                }
    };

    public class ItemsData
    {
        public int Id { get; set; }
        public string Name { get; set; }
        //  Skip many other properties
 
        public Dictionary<string, string> Dynamics{ get; set; }
    }

```

The expected Json is `{ "id": 1, "name": "k", ... , "prop1":"1", "prop2":"2", "prop3":"3" } `. 

It's easy to be done with a specific `JsonConverter`. But if there're many similar classes, and each class has many other properties except  `Dynamics`, I have to write a lot of `JsonConverter` and each of them has long lines of codes in `Write`.

```csharp
        public override void Write(Utf8JsonWriter writer, ItemsData value, JsonSerializerOptions options)
        {
            var namingPolicy = options?.PropertyNamingPolicy;

            writer.WriteStartObject();

            writer.WriteNumber(GetPropName(nameof(value.Id), namingPolicy), value.Id);
            writer.WriteString(GetPropName(nameof(value.Name), namingPolicy), value. Name);
           // Skip a lot of lines

            if (value.Dynamics?.Count > 0)
            {
                foreach (var item in value. Dynamics)
                {
                    writer.WriteString(item.Key, item. Value);
                }
            }

            writer.WriteEndObject();
        }

        private static string GetPropName(string name, JsonNamingPolicy policy)
        {
            return policy == null
                ? name
                : policy.ConvertName(name);
        }
```

Any better solutions? Thanks in advance.

**Comment by pinkfloydx33** [comment, marked as answer]
Does [`[JsonExtensionData]`](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.serialization.jsonextensiondataattribute?view=net-6.0) on the dictionary property not work for you? A potential downside is that you'd have to change the dictionary's value type to either `object` or `JsonElement`.

**Reply by LeaFrock** [reply]
Ahh... that's what I want, and it's documented in [an unremarkable doc](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/handle-overflow). Thanks!

**Reply by steveharter** [reply]
Note that a `JsonObject` property can also be specified for `[JsonExtensionData]`. Since `JsonObject` is editable, this allows the extension data to be modified if that is a scenario.
